### PR TITLE
[nrf noup] fs: littlefs: Add littlefs to pm.yml to support the littlefs backend

### PR DESCRIPTION
**Pull request by sigvartmh**
Add `#ifdef` for littlefs so that littlefs can be used in the same way
as NFFS with pm.yml configurations.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

File: subsys/fs/Kconfig [hunk 1]
```diff
@@ -83,6 +83,13 @@ config PM_PARTITION_SIZE_NFFS_STORAGE
 	help
 	  Flash space set aside for the nffs storage partition.
 
+config PM_PARTITION_SIZE_LITTLEFS
+	hex "Flash space reserved for nffs storage partition"
+	default 0x6000
+	help
+	  Flash space set aside for the nffs storage partition.
+# end nrf/subsys/partition_manager/Kconfig.template.partition_size
```
Comment (hakonfam) [inline subsys/fs/Kconfig]: What is this?

File: subsys/fs/Kconfig [hunk 2]
```diff
@@ -83,6 +83,13 @@ config PM_PARTITION_SIZE_NFFS_STORAGE
 	help
 	  Flash space set aside for the nffs storage partition.
 
+config PM_PARTITION_SIZE_LITTLEFS
+	hex "Flash space reserved for nffs storage partition"
```
Comment (hakonfam) [inline subsys/fs/Kconfig]:
```suggestion
	hex "Flash space reserved for littlefl storage partition"
```

File: subsys/fs/Kconfig (hunk 1 above)
Comment (sigvartmh) [reply subsys/fs/Kconfig]: #end clause :) but I can see that there is a copy-paste error ther

File: subsys/fs/pm.yml [hunk 3]
```diff
@@ -5,3 +5,9 @@ nffs_storage:
   placement: {after: [mcuboot_storage, app]}
   size: CONFIG_PM_PARTITION_SIZE_NFFS_STORAGE
 #endif
+
+#ifdef CONFIG_FILE_SYSTEM_LITTLEFS
+littlefs:
+  placement: {after: [mcuboot_storage, app]}
```
Comment (hakonfam) [inline subsys/fs/pm.yml]:
```suggestion
  placement: {after: [nffs_storage, mcuboot_storage, app]}
```

Comment (sigvartmh) [reply subsys/fs/pm.yml]: It's a bit weird to use both file-systems but I guess it has a use-case.

File: subsys/fs/Kconfig (hunk 1 above)
Comment (sigvartmh) [reply subsys/fs/Kconfig]: removed as it was moved to a different place

**Comment by hakonfam** [comment]
Could you verify the change by running some settings test on pca10090ns with this backend? (If supported)

**Comment by sigvartmh** [comment]
Fails on some of the littlefs tests with nrf9160 . I'll try with nrf52840 which boats as white-listed board I think some configuration is wrong for the test.

**Comment by mbolivar** [comment]
History in this repository has been rewritten. Please rebase this PR.

**Comment by mbolivar** [comment]
> fs: littlefs: Add littlefs to pm.yml to support the littlefs backend

needs `[nrf noup]` sauce tag

**Comment by sigvartmh** [comment]
Added but this is untested. I'll come back to this PR after I'm done with my priorities :smile:

File: subsys/fs/pm.yml (hunk 3 above)
Comment (nvlsianpu) [reply subsys/fs/pm.yml]: not any more.

**Comment by sigvartmh** [comment]
After trying to get it to run with the tests. I got it to run with only the smallest partition enabled. It passes the test suits but there are some errors reported.

```
Running test suite littlefs_test
===================================================================
starting test - test_util_path_init_base
PASS - test_util_path_init_base
===================================================================
starting test - test_util_path_init_overrun
PASS - test_util_path_init_overrun
===================================================================
starting test - test_util_path_init_extended
PASS - test_util_path_init_extended
===================================================================
starting test - test_util_path_extend
PASS - test_util_path_extend
===================================================================
starting test - test_util_path_extend_up
PASS - test_util_path_extend_up
===================================================================
starting test - test_util_path_extend_overrun
PASS - test_util_path_extend_overrun
===================================================================
starting test - test_lfs_basic
clearing partition /sml
Erasing 65536 (0x10000) bytes
Wiped flash area 2 for /sml
mounting /sml
I: LittleFS version 2.1, disk version 2.0
I: FS at NRF_FLASH_DRV_NAME:0xf0000 is 16 0x1000-byte blocks with 512 cycle
I: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
E: Corrupted dir pair at 0 1
W: can't mount (LFS -84); formatting
I: /sml mounted
checking clean statvfs of /sml
/sml: bsize 16 ; frsize 4096 ; blocks 16 ; bfree 14
creating and writing file
opening and verifying file
seek and tell in file
truncate in file
unlink hello
E: failed get file or dir stat (-2)
sync goodbye
sync size 64
unmounting /sml
I: /sml unmounted
checking double unmount diagnoses
E: fs ops functions not set!!
mounting /sml
I: LittleFS version 2.1, disk version 2.0
I: FS at NRF_FLASH_DRV_NAME:0xf0000 is 16 0x1000-byte blocks with 512 cycle
I: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
I: /sml mounted
verify goodbye
I: /sml unmounted
PASS - test_lfs_basic
===================================================================
starting test - test_lfs_dirops
checking clean mount
Erasing 65536 (0x10000) bytes
Wiped flash area 2 for /sml
I: LittleFS version 2.1, disk version 2.0
I: FS at NRF_FLASH_DRV_NAME:0xf0000 is 16 0x1000-byte blocks with 512 cycle
I: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
E: Corrupted dir pair at 0 1
W: can't mount (LFS -84); formatting
I: /sml mounted
checking dir create unlink
E: failed to unlink path (-15)
building layout on /sml
before: bsize 16 ; frsize 4096 ; blocks 16 ; bfree 14
create at /sml/f1 with 1 from 0x01: 0
create at /sml/f2 with 100 from 0x02: 0
mkdir /sml/d1: 0
create at /sml/d1/d1f1 with 23 from 0x0b: 0
create at /sml/d1/d1f2 with 612 from 0x0c: 0
exit directory /sml/d1
create at /sml/f3 with 10000 from 0x03: 0
after: bsize 16 ; frsize 4096 ; blocks 16 ; bfree 7
checking layout
check /sml for 7 entries
/sml d1/ 0
check /sml/d1 for 2 entries
/sml/d1 d1f1 23
/sml/d1 d1f2 612
/sml/d1 found 2 entries, 0 foreign
/sml f1 1
/sml f2 100
/sml f3 10000
/sml found 4 entries, 0 foreign
verifying f1 1
verifying f2 100
verifying d1/ 0
verifying d1f1 23
verifying d1f2 612
verifying f3 10000
create at /sml/rename/f1f with 8 from 0x01: 0
create at /sml/rename/f2f with 8 from 0x02: 0
create at /sml/rename/f2t with 8 from 0x03: 0
create at /sml/rename/f3f with 8 from 0x04: 0
create at /sml/rename/f4f with 8 from 0x05: 0
mkdir /sml/rename/d1f: 0
create at /sml/rename/d1f/d1f1 with 16 from 0x05: 0
exit directory /sml/rename/d1f
mkdir /sml/rename/d2t: 0
create at /sml/rename/d2t/d2f1 with 16 from 0x06: 0
exit directory /sml/rename/d2t
check /sml/rename for 11 entries
/sml/rename d1f/ 0
check /sml/rename/d1f for 1 entries
/sml/rename/d1f d1f1 16
/sml/rename/d1f found 1 entries, 0 foreign
/sml/rename d2t/ 0
check /sml/rename/d2t for 1 entries
/sml/rename/d2t d2f1 16
/sml/rename/d2t found 1 entries, 0 foreign
/sml/rename f1f 8
/sml/rename f2f 8
/sml/rename f2t 8
/sml/rename f3f 8
/sml/rename f4f 8
/sml/rename found 7 entries, 0 foreign
/sml/rename/nofile => /sml/rename/f1t -ENOENT
E: failed to rename file or dir (-2)
/sml/rename/f1f => /sml/rename/f1t ok
/sml/rename/f2f => /sml/rename/f2t clobber ok
E: failed get file or dir stat (-2)
/sml/rename/f3f => /sml/rename/d1f/d1f2t move ok
E: failed get file or dir stat (-2)
/sml/rename/d1f => /sml/rename/d2t -ENOTEMPTY
E: failed to rename file or dir (-15)
/sml/rename/d1f => /sml/rename/d1t ok
E: failed get file or dir stat (-2)
check /sml/rename for 10 entries
/sml/rename d1t/ 0
check /sml/rename/d1t for 2 entries
/sml/rename/d1t d1f1 16
/sml/rename/d1t d1f2t 8
/sml/rename/d1t found 2 entries, 0 foreign
/sml/rename d2t/ 0
check /sml/rename/d2t for 1 entries
/sml/rename/d2t d2f1 16
/sml/rename/d2t found 1 entries, 0 foreign
/sml/rename f1t 8
/sml/rename f2t 8
/sml/rename f4f 8
/sml/rename found 5 entries, 0 foreign
I: /sml unmounted
PASS - test_lfs_dirops
===================================================================
starting test - test_lfs_perf
clearing /sml for small 8x1K dflt write/read test
Erasing 65536 (0x10000) bytes
Wiped flash area 2 for /sml
I: LittleFS version 2.1, disk version 2.0
I: FS at NRF_FLASH_DRV_NAME:0xf0000 is 16 0x1000-byte blocks with 512 cycle
I: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
E: Corrupted dir pair at 0 1
W: can't mount (LFS -84); formatting
I: /sml mounted
/sml: bsize 16 ; frsize 4096 ; blocks 16 ; bfree 14
read_size 16 ; prog_size 16 ; cache_size 64 ; lookahead_size 32
creating and writing 8 1024-byte blocks
small 8x1K dflt write 8 * 1024 = 8192 bytes in 372 ms: 22021 By/s, 21 KiBy/s
small 8x1K dflt read 8 * 1024 = 8192 bytes in 1 ms: 8192000 By/s, 8000 KiBy/s
I: /sml unmounted
bufs 0x20023840 0x20024040 0x20024840
clearing /sml for small 8x1K bigfile write/read test
Erasing 65536 (0x10000) bytes
Wiped flash area 2 for /sml
I: LittleFS version 2.1, disk version 2.0
I: FS at NRF_FLASH_DRV_NAME:0xf0000 is 16 0x1000-byte blocks with 512 cycle
I: sizes: rd 256 ; pr 256 ; ca 1024 ; la 128
E: Corrupted dir pair at 0 1
W: can't mount (LFS -84); formatting
I: /sml mounted
/sml: bsize 256 ; frsize 4096 ; blocks 16 ; bfree 14
read_size 256 ; prog_size 256 ; cache_size 1024 ; lookahead_size 128
creating and writing 8 1024-byte blocks
small 8x1K bigfile write 8 * 1024 = 8192 bytes in 367 ms: 22321 By/s, 21 KiBy/s
small 8x1K bigfile read 8 * 1024 = 8192 bytes in 1 ms: 8192000 By/s, 8000 KiBy/s
I: /sml unmounted
PASS - test_lfs_perf
===================================================================
Test suite littlefs_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by sigvartmh** [comment]
Added test modifications and board specific `prj.conf` for completeness

**Comment by mbolivar** [comment]
This repository's history has been rewritten for the v1.2 release. Please rebase this PR if still relevant, or close it if no longer needed. Thanks!

**Comment by nvlsianpu** [comment]
DNM until Upmerge #296 will be merged

**Comment by joerchan** [comment]
More than a year old now, does it still apply?

Needs to be rebased after 1.4.0 history rewrite.

**Comment by sigvartmh** [comment]
Not sure if it's still needed.

**Comment by hakonfam** [comment]
The majority of this code should be in the nrf repo now anyway. Also the flash_map.h changes should make this obsolete. Closing.

**Pull request closed without merging**